File: srei-main-funcional/srei/src/main/java/srei/proyecto/srei/usuario/UsuarioService.java
package srei.proyecto.srei.usuario;

import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import srei.proyecto.srei.usuario.dto.RolDto;
import srei.proyecto.srei.usuario.dto.UsuarioCrearRequest;
import srei.proyecto.srei.usuario.dto.UsuarioDto;

import java.util.List;
import java.util.Set;

@Service
@RequiredArgsConstructor
public class UsuarioService {

    private final UsuarioRepository usuarioRepository;
    private final RolRepository rolRepository;
    private final PasswordEncoder encoder;

    public List<UsuarioDto> listar() {
        return usuarioRepository.findAll().stream().map(this::toDto).toList();
    }

    public UsuarioDto crear(UsuarioCrearRequest req) {
        if (usuarioRepository.existsByCorreoIgnoreCase(req.correo())) {
            throw new IllegalArgumentException("El correo ya está registrado");
        }

        RolNombre rolNombre = req.rol() == null ? RolNombre.ASISTENTE : req.rol();
        Rol rol = rolRepository.findByNombre(rolNombre).orElseThrow();

        Usuario u = Usuario.builder()
                .nombres(req.nombres())
                .apellidos(req.apellidos())
                .correo(req.correo())
                .contrasenaHash(encoder.encode(req.contrasena()))
                .habilitado(true)
                .facultad(req.facultad())
                .carrera(req.carrera())
                .roles(Set.of(rol))
                .build();

        return toDto(usuarioRepository.save(u));
    }

    public UsuarioDto setHabilitado(Long id, boolean valor) {
        Usuario u = usuarioRepository.findById(id).orElseThrow(() -> new IllegalArgumentException("Usuario no existe"));
        u.setHabilitado(valor);
        return toDto(usuarioRepository.save(u));
    }

    public UsuarioDto asignarRol(Long id, RolNombre rolNombre) {
        Usuario u = usuarioRepository.findById(id).orElseThrow(() -> new IllegalArgumentException("Usuario no existe"));
        Rol rol = rolRepository.findByNombre(rolNombre).orElseThrow(() -> new IllegalArgumentException("Rol no existe"));
        u.getRoles().add(rol);
        return toDto(usuarioRepository.save(u));
    }

    private UsuarioDto toDto(Usuario u) {
        return new UsuarioDto(
                u.getId(),
                u.getNombres(),
                u.getApellidos(),
                u.getCorreo(),
                u.isHabilitado(),
                u.getFacultad(),
                u.getCarrera(),
                u.getRoles().stream().map(r -> r.getNombre().name()).toList()
        );
    }

    public List<RolDto> listarRol() {
        return rolRepository.findAll().stream().map(this::RtoDto).toList();
    }
    
    private RolDto RtoDto(Rol r) {
        return new RolDto(
                r.getId(),
                r.getNombre()
        );
    }

}

File: srei-main-funcional/frontend/src/app/core/services/usuario.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../../environments/environment';

export interface UsuarioDto {
  id: number;
  nombres: string;
  apellidos: string;
  correo: string;
  habilitado: boolean;
  facultad?: string;
  carrera?: string;
  roles: string[];
}
export interface RolDto {
  id: number;
  nombre: string;
}

@Injectable({ providedIn: 'root' })
export class UsuarioService {
  private apiUrl = environment.apiUrl;
  constructor(private http: HttpClient) {}

  listar() {
    return this.http.get<UsuarioDto[]>(this.apiUrl + '/usuarios');
  }

  crear(payload: any) {
    return this.http.post<UsuarioDto>(this.apiUrl + '/usuarios', payload);
  }

  habilitar(id: number, valor: boolean) {
    return this.http.patch<UsuarioDto>(this.apiUrl + `/usuarios/${id}/habilitado?valor=${valor}`, {});
  }
  actualizarRol(id: number, rol: string) {
  return this.http.put(
    this.apiUrl + '/usuarios/' + id + '/roles',
    { rol }
  );
  
}
 listarRoles(){
     return this.http.get<RolDto[]>(
    this.apiUrl + '/usuarios/roles'
  );
   }
}
